fleetspeak/src/server/spanner/messagestore.go

```diff
@@ -0,0 +1,778 @@
+		if err != nil {
+			return nil, mus, err
+		}
+		mus = append(mus, mu)
+		msgKeySet = spanner.KeySets(
+			spanner.KeySets(spanner.Key{pm.MessageID}), msgKeySet)
+		count++
+	}
+
+	rl := make([]*fspb.Message, 0, count)
+
+	txn2 := d.dbClient.Single()
+	defer txn2.Close()
+	iter = txn2.Read(ctx, d.messages, msgKeySet,
+		[]string{"MessageID", "Source", "SourceMessageID", "Destination", "MessageType", "CreationTime",
+			"EncryptedData", "Result", "ValidationInformation", "Annotations"})
+	defer iter.Stop()
+	for {
+		row, err := iter.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, mus, err
+		}
+		var messageId []byte
+		var source *fspb.Address
+		var sourceMessageId []byte
+		var destination *fspb.Address
+		var messageType string
+		var creationTime time.Time
+		encryptedData := &spanner.NullProtoMessage{}
+		encryptedData.ProtoMessageVal = &anypb.Any{}
+		result := &spanner.NullProtoMessage{}
+		result.ProtoMessageVal = &fspb.MessageResult{}
+		validationInfo := &spanner.NullProtoMessage{}
+		validationInfo.ProtoMessageVal = &fspb.ValidationInfo{}
+		annotations := &spanner.NullProtoMessage{}
+		annotations.ProtoMessageVal = &fspb.Annotations{}
+		if err := row.Columns(&messageId, &source, &sourceMessageId, &destination, &messageType, &creationTime, &encryptedData, &result, &validationInfo, &annotations); err != nil {
+			return nil, mus, err
+		}
+		m := message{}
+		m.MessageID = messageId
+		m.Source = source
+		m.SourceMessageID = sourceMessageId
+		m.Destination = destination
+		m.MessageType = messageType
+		m.CreationTime = creationTime
+		if encryptedData.Valid {
+			m.EncryptedData = encryptedData.ProtoMessageVal.(*anypb.Any)
+		}
+		if result.Valid {
+			m.Result = result.ProtoMessageVal.(*fspb.MessageResult)
+		}
+		if validationInfo.Valid {
+			m.ValidationInformation = validationInfo.ProtoMessageVal.(*fspb.ValidationInfo)
+		}
+		if annotations.Valid {
+			m.Annotations = annotations.ProtoMessageVal.(*fspb.Annotations)
+		}
+
+		log.V(2).InfoContextf(ctx, "====== messagestore: Message - MessageID: %v", m.MessageID)
+		rl = append(rl, protoMessage(&m))
+	}
+
+	return rl, mus, nil
+}
+
+// RegisterMessageProcessor implements db.Store.
+func (d *Datastore) RegisterMessageProcessor(mp db.MessageProcessor) {
+	ctx := context.Background()
+	go func() {
+		err := d.pubsubSub.Receive(ctx, func(_ context.Context, msg *pubsub.Message) {
+			var msgKeySet = spanner.KeySets()
+			msgKeySet = spanner.KeySets(
+				spanner.KeySets(spanner.Key{msg.Data}), msgKeySet)
+			msgs, err := d.tryGetMessages(ctx, msgKeySet, true)
+			if err == nil {
+				msg.Ack()
+				mp.ProcessMessages(msgs)
+			} else {
+				msg.Nack()
+			}
+		})
+		if err != nil && err != context.Canceled {
+			log.Errorf("Failed to receive server message for processing: %v", err)
+		}
+	}()
+}
+
+func (d *Datastore) StopMessageProcessor() {
+	log.Error("+++ messagestore: StopMessageProcessor() called")
```

Comment:
```suggestion
func (d *Datastore) StopMessageProcessor() {
        d.pubsubTopic.Stop()
	log.Error("+++ messagestore: StopMessageProcessor() called")
```
[`Stop`](https://pkg.go.dev/cloud.google.com/go/pubsub#Topic.Stop) will flush all outstanding messages and will clean up the remaining go-routines.